### ISSUE TITLE
error when collecting tests, db not created

### ISSUE THREAD

**Issue by aljosa** [pytest-dev/pytest-django#202]
i'm running into a strange issue where locally (on at least 2 developer machines) tests are passing as expected but fail when run on circleci.com.

when pytest is collecting tests a part of code which runs db query is also executed.
this doesn't seem to be a problem except when run in circleci environment.
versions are same locally and on circleci.

don't know if it makes any difference, here is part of error when run on circleci:
```
make test
make: *** [test] Error 1
py.test --create-db tests/
Test session starts (platform: linux2, Python 2.7.8, pytest 2.6.4, pytest-sugar 0.3.5)
plugins: cov, django, sugar, xdist, bdd, splinter


――――――――――――――― ERROR collecting tests/[ ... ].py ―――――――――――――――
tests/[ ... ].py:15: in <module>
[ ... ]
../virtualenvs/venv-2.7.8/lib/python2.7/site-packages/django/db/models/query.py:96: in __iter__
    self._fetch_all()
../virtualenvs/venv-2.7.8/lib/python2.7/site-packages/django/db/models/query.py:857: in _fetch_all
    self._result_cache = list(self.iterator())
../virtualenvs/venv-2.7.8/lib/python2.7/site-packages/django/db/models/query.py:1068: in iterator
    for row in self.query.get_compiler(self.db).results_iter():
../virtualenvs/venv-2.7.8/lib/python2.7/site-packages/django/db/models/sql/compiler.py:713: in results_iter
    for rows in self.execute_sql(MULTI):
../virtualenvs/venv-2.7.8/lib/python2.7/site-packages/django/db/models/sql/compiler.py:786: in execute_sql
    cursor.execute(sql, params)
../virtualenvs/venv-2.7.8/lib/python2.7/site-packages/django/db/backends/util.py:53: in execute
    return self.cursor.execute(sql, params)
../virtualenvs/venv-2.7.8/lib/python2.7/site-packages/django/db/utils.py:99: in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
../virtualenvs/venv-2.7.8/lib/python2.7/site-packages/django/db/backends/util.py:53: in execute
    return self.cursor.execute(sql, params)
E   ProgrammingError: relation "[ TABLE_NAME ]" does not exist
E   LINE 1: SELECT "TABLE_NAME"."FIELD" FROM "TABLE_NAME" WHERE ("...
E                                               ^
```

**Comment by aljosa** [comment]
i know this looks like an environment issue but i'm hoping somebody had similar problems and can share some info

**Comment by pelme** [comment]
What happens on line 15 in `tests/[ ... ].py`? It looks like something directly in the module tries to issue a database query (i.e. at test import time).

That would explain that it works on a developer machine since it probably connects to the real database (which exists) and issues the query against that db instead.

We could probably to improve this in pytest-django by disabling the database cursor earlier and raise an explicit error. Then this would also make it explicitly fail on the developer machines.

**Comment by aljosa** [comment]
@pelme yes, i mentioned this in first comment. code runs db query when pytest is collecting tests.

running "py.test --create-db tests/" w/ same requirements installed on local machine and on circleci gives different results, which is a serious problem for me.

is there a way i can trigger "create db" (or whatever happens) before py.test starts collecting tests?

i understand that a valid suggestion is to fix this by removing code which does db query but i would like to avoid adding a few hundred lines of code on a project which heavily relies on db just to create environment where pytest can collect tests before db is created.
creating test db before pytest starts collecting tests would be the best option for me, is it possible to do that?

**Comment by pelme** [comment]
This is not intended to work locally either - it is buggy locally too, just in another way. The ordinary Django database is accessed when you run tests locally right now, not the test database.

The pytest-django issue here is that it *doesn't* fail locally, I would like to fix that.

The reason for this is that the database is managed with pytest fixtures, and doing database modifications outside of the fixtures leaves not control over the database to pytest-django, therefore it is not really possible to support this. Fixture activation happens after test collection (test and fixture definitions must be collected before they can be executed).

A potential solution would be:
```python

@pytest.fixture(scope='session', autouse=True)
def _do_database_stuff(_django_db_setup):
    pass # do everything that currently happens in the module
```

Does that work?

**Comment by pelme** [comment]
Or maybe this gist can be useful, you would need to enable the database cursor too:
https://gist.github.com/magopian/7673776

**Comment by aljosa** [comment]
thanks for suggestions
